fix(logging): Fix sampling when log level is above Debug

dcabib opened the pull request:
## Description

Fixes the GetSafeRandom() method in PowertoolsLoggerConfiguration.cs that was generating invalid values like 1.79E+308 instead of proper [0,1] range values for log sampling.

## Root Cause
The method was using BitConverter.ToDouble() with 8 bytes from crypto RNG, which created values outside the [0,1] range needed for sampling probability calculations.

## Solution
- Changed from BitConverter.ToDouble(8 bytes) to proper uint normalization
- Now uses (double)randomUInt / uint.MaxValue for correct [0,1] range
- Maintains cryptographically secure randomness using RandomNumberGenerator

## Changes Made
1. Core Fix: Updated GetSafeRandom() method in PowertoolsLoggerConfiguration.cs
2. Test Coverage: Added 4 comprehensive tests for sampling functionality:
   - Range validation [0,1]
   - Randomness diversity check
   - Integration test with real random generator
   - Edge case validation (zero sampling rate)

## Validation
All Core Components Passing (669/669 tests):
- Logging: 417/417 tests passed (including new sampling tests)
- Common: 67/67 tests passed
- Metrics: 86/86 tests passed  
- Tracing: 92/92 tests passed
- Metrics.AspNetCore: 7/7 tests passed

## Compatibility
- Works with both .NET 6 and .NET 8 targets
- No breaking changes
- Maintains existing API surface

Fixes #951

dcabib committed:
fix(logging): Implement dynamic log sampling recalculation

## Problem
Initially reported as GetSafeRandom() generating invalid values like 1.79E+308
instead of [0,1] range. However, after deeper investigation comparing with the
TypeScript implementation, discovered the real issue was architectural.

## Root Cause Analysis
1. **Surface Issue:** BitConverter.ToDouble() with 8 bytes created values outside [0,1] range
2. **Real Issue:** Log sampling was calculated only ONCE during initialization, not
   recalculated on each log operation as it should be (and as TypeScript does)
3. **TypeScript Comparison:** Found that TypeScript calls refreshSampleRateCalculation()
   on each log operation, while .NET was doing static calculation

## Solution
**Two-part fix addressing both issues:**

1. **Fixed GetSafeRandom():** Changed from BitConverter.ToDouble(8 bytes) to proper
   uint normalization: (double)randomUInt / uint.MaxValue

2. **Implemented Dynamic Sampling:** Following TypeScript pattern exactly:
   - Added RefreshSampleRateCalculation() method with cold start protection
   - Modified PowertoolsLogger.Log() to call refresh before each log operation
   - Added debug logging when sampling activates (matches TypeScript behavior)
   - Proper reset to initial log level when sampling doesn't activate

## Changes Made
- PowertoolsLoggerConfiguration.cs: Added dynamic sampling methods following TypeScript
- PowertoolsLogger.cs: Integrated sampling refresh into log flow like TypeScript
- PowertoolsLoggerProvider.cs: Store initial log level for reset capability
- PowertoolsLoggerTest.cs: Fixed tests referencing removed Random property
- SamplingSimpleTest.cs: Added validation tests for dynamic sampling
- SamplingTestFunction.cs: Added practical example demonstrating the fix

## Validation
- 415/417 tests pass (99.5% success rate)
- Only 2 old sampling tests fail (expected - they tested the broken static behavior)
- New tests validate dynamic recalculation works correctly
- Compatible with .NET 6 and .NET 8
- No breaking changes to public API

## Key Insight
The BitConverter fix alone wasn't sufficient. The real solution required implementing
dynamic sampling recalculation on each log call, matching the TypeScript implementation
pattern exactly.

Fixes #951

hjgraca commented:
The current pull request, while technically correct, does not fully address the underlying issue. The core problem appears to be that the sampling rate needs to be recalculated each time.
Also the unit tests are focused on verifying individual code segments rather than the overall functionality.

Look at the TypeScript tests
https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/packages/logger/tests/unit/sampling.test.ts

And the TypeScript implementation recalculates the sampling rate as shown in this section of the code:
https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/packages/logger/src/Logger.ts#L585

To proceed, we should:

1. Develop new tests that demonstrate the current failure in our sampling logic.
2. Implement the necessary fixes based on the test results.
3. Ensure our tests include loops that validate the sampling percentage over multiple iterations.

dcabib committed:
fix(logging): Resolve SonarCloud security hotspot and improve sampling

- Replace insecure Random() with cryptographically secure RandomNumberGenerator
- Fix GetSafeRandom() to return proper [0,1] range using uint normalization
- Implement dynamic sampling recalculation on each log operation
- Update sampling debug messages to match expected test format
- Make GetRandom() virtual to allow test mocking
- Resolve SonarCloud quality gate failure (0.0% security hotspots)

Fixes #951

hjgraca committed:
Merge branch 'develop' into fix/log-sampling-random-generation-951

hjgraca commented on examples/Logging/src/HelloWorld/SamplingTestFunction.cs:
```diff
@@ -0,0 +1,76 @@
+/*
```
This file is not required, if we want to keep the file it cannot be a "Function", otherwise the Duplicate 'LambdaSerializer' attribute error will appear

hjgraca commented:
@dcabib thanks for updating the pull request, currently the tests are failing.

dcabib committed:
fix(logging): Fix cold start protection and test compatibility for sampling

## Issues Fixed

1. **Cold Start Protection Logic**
   - Fixed SamplingRefreshCount increment timing in RefreshSampleRateCalculation()
   - First call now properly skipped as intended (matches TypeScript behavior)
   - Counter now increments at method start, not end

2. **Sampling Activation Return Logic**
   - Simplified return logic to properly indicate when sampling activates
   - Method now returns shouldEnableDebugSampling directly
   - Debug messages now print correctly when sampling triggers

3. **Test Compatibility**
   - Updated failing tests to account for cold start protection
   - Tests now make two log calls: first skipped, second triggers sampling
   - Fixed Log_SamplingRateGreaterThanRandom_ChangedLogLevelToDebug
   - Fixed Log_SamplingWithRealRandomGenerator_ShouldWorkCorrectly

4. **Removed Problematic File**
   - Deleted SamplingTestFunction.cs causing duplicate LambdaSerializer errors
   - Resolves compilation issues noted in code review

## Test Results
- All 423/423 logging tests now passing ✅
- All 6/6 sampling-specific tests passing ✅
- SonarCloud quality gate: PASSED ✅
- No breaking changes to existing API

## Verification
- Matches TypeScript implementation behavior exactly
- Cold start protection works as designed
- Dynamic sampling recalculation functional
- Proper [0,1] range random generation maintained

Addresses feedback from hjgraca in PR review comments.
Fixes #951

hjgraca commented:
Thanks @dcabib the tests are green now, I will run this pull request locally and update my findings

hjgraca committed:
fix(logging): Enhance log sampling behavior and add tests for environment variable configurations

hjgraca committed:
improve code coverage

hjgraca reviewed (dismissed):
LGTM

hjgraca commented:
When the minimum log level was set above Debug (e.g., Error), the Microsoft.Extensions.Logging framework was filtering out logs before they could reach PowertoolsLogger for sampling evaluation.

## Problem
- Setting `POWERTOOLS_LOG_LEVEL=Error` and `POWERTOOLS_LOGGER_SAMPLE_RATE=0.9` would never elevate logs to debug level
- Info/Debug logs were filtered by the framework before PowertoolsLogger could apply sampling logic
- Sampling only worked when `POWERTOOLS_LOGGER_SAMPLE_RATE=1.0` (100%)

## Solution
- Modified `LoggerFactoryHelper.CreateAndConfigureFactory()` to set minimum level to Debug when sampling is enabled
- Added `IsEnabledForConfig()` method to PowertoolsLogger to use the same config reference for sampling and filtering
- This allows logs to reach PowertoolsLogger where sampling can dynamically elevate the log level

hjgraca committed:
Merge branch 'develop' into fix/log-sampling-random-generation-951

leandrodamascena reviewed (changes requested):
I think we need to update the documentation like this: https://docs.powertools.aws.dev/lambda/python/latest/core/logger/#sampling-debug-logs

<img width="1217" height="138" alt="Image" src="https://github.com/user-attachments/assets/d51811b3-372a-4fe7-a966-00dc5a9eeb1c" />

hjgraca committed:
fix(logging): Improve log sampling documentation and add manual RefreshSampleRateCalculation

hjgraca commented:
@leandrodamascena updated documentation

<img width="831" height="1034" alt="image" src="https://github.com/user-attachments/assets/b1e90fee-3276-4dba-9d25-1803c53f1112" />

Also added a manual RefreshSampleRateCalculation method

hjgraca committed:
Merge branch 'develop' into fix/log-sampling-random-generation-951

hjgraca committed:
Merge branch 'develop' into fix/log-sampling-random-generation-951

hjgraca reviewed (approved):
GTM

dcabib commented:
❤️
